library/std/src/io/buffered/bufwriter.rs

```diff
@@ -287,67 +301,173 @@ impl<W: Write> BufWriter<W> {
             self.panicked = false;
             r
         } else {
-            self.buf.extend_from_slice(buf);
+            // Write to the buffer. In this case, we write to the buffer even if it fills it
+            // exactly. Doing otherwise would mean flushing the buffer, then writing this
+            // input to the inner writer, which in many cases would be a worse strategy.
+
+            // SAFETY: We just called `self.flush_buf()`, so `self.buf.len()` is 0, and
+            // we entered this else block because `buf.len() < self.buf.capacity()`.
+            // Therefore, `buf.len() <= self.buf.capacity() - self.buf.len()`.
+            unsafe {
+                self.write_to_buffer_unchecked(buf);
+            }
+
             Ok(buf.len())
         }
     }
 
-    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+    // If this function ends up being called frequently relative to `write_all`,
+    // it's likely a sign that the client is using an improperly sized buffer
+    // or their write patterns are somewhat pathological.
+    #[cold]
+    fn write_all_cold(&mut self, buf: &[u8]) -> io::Result<()> {
         // Normally, `write_all` just calls `write` in a loop. We can do better
         // by calling `self.get_mut().write_all()` directly, which avoids
         // round trips through the buffer in the event of a series of partial
         // writes in some circumstances.
-        if self.buf.len() + buf.len() > self.buf.capacity() {
+
+        if buf.len() > self.spare_capacity() {
             self.flush_buf()?;
         }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
+
+        // Why not len > capacity? To avoid a needless trip through the buffer when the input
+        // exactly fills it. We'd just need to flush it to the underlying writer anyway.
         if buf.len() >= self.buf.capacity() {
             self.panicked = true;
             let r = self.get_mut().write_all(buf);
             self.panicked = false;
             r
         } else {
-            self.buf.extend_from_slice(buf);
+            // Write to the buffer. In this case, we write to the buffer even if it fills it
+            // exactly. Doing otherwise would mean flushing the buffer, then writing this
+            // input to the inner writer, which in many cases would be a worse strategy.
+
+            // SAFETY: We just called `self.flush_buf()`, so `self.buf.len()` is 0, and
+            // we entered this else block because `buf.len() < self.buf.capacity()`.
+            // Therefore, `buf.len() <= self.buf.capacity() - self.buf.len()`.
+            unsafe {
+                self.write_to_buffer_unchecked(buf);
+            }
+
+            Ok(())
+        }
+    }
+
+    // SAFETY: Requires `buf.len() <= self.buf.capacity() - self.buf.len()`,
+    // i.e., that input buffer length is less than or equal to spare capacity.
+    #[inline]
+    unsafe fn write_to_buffer_unchecked(&mut self, buf: &[u8]) {
```

Comment:
Minor nitpick: for private functions, the unadorned `inline` attribute is redundant: the compiler always has access to the function body and is therefore able to make inlining decisions.

Reply:
But doesn't this encourage inlining slightly? It adds the `inlinehint` attribute to the LLVM IR anyway.

And if there's a question of whether or not we want to encourage inlining, I think we really do, as the point of the optimization is to inline writes to the internal buffer in hot code. I only separated it out into a function for code reuse. The `#[inline]` is just saying "please don't undo the optimization I'm trying to do". :)

Of course, we could trust the compiler to make the right decision based on context. There are probably cases where `write` and `write_all` are called in cold code where the client doesn't want this function to be inlined. But I'm hesitant to optimize for that case at the risk of undoing the intended optimization. Also, in those cases, I think they actually wouldn't want `write` or `write_all` to be inlined, and would need to do something like wrap them in a function marked `#[inline(never)]` anyway.